### PR TITLE
Address 7.x-ISLANDORA-2170.

### PR DESCRIPTION
**Pull request by mjordan**
**JIRA Ticket**: (https://jira.duraspace.org/browse/ISLANDORA-2170)

# What does this Pull Request do?

Defines hooks to notify modules that the given object, collection, content model access event was persisted, and that a datastream download event was persisted.

# What's new?

Fires the new hooks immediately after the functions that persist the object access, collection access, content model access, and datastream download are fired. Passes into the hook the relevant data to allow for useful logic within hook implementations.

Also, `islandora_usage_stats_object_view_to_db()` currently returns the id of the row added to record the object access, but the corresponding functions for content model, collection, and datastream downloads don't:

* `islandora_usage_stats_content_model_view_to_db()`
* `islandora_usage_stats_collection_view_to_db()`
* `islandora_usage_stats_datastream_download_to_db()`

I have changed these functions so they now do return their respective row IDs. Since the return value of these functions was never checked (it would have been `false`), there should be no side effects from adding a return value.

# How should this be tested?

1. At `admin/islandora/tools/islandora_usage_stats`, make sure "Omit IPs" is emtpy and "Cooldown time" is 0.
1. Install and enable https://github.com/mjordan/islandora_2170_test, which implements all of the new hooks. This module adds a `dsm()` that shows the values available for object, collection, and content model access events For datastream access, it doesn't fire `dsm()`, it fires `dd()`.
1. After you have enabled the test module, visit a non-collection object page. You should see three messages, one for each of object, collection, and content model access events.
  * If you view a basic image object, you will see a fourth dsm message that only displays for basic images.
1. To test datastream downloads, attempt to download the object's TN datastream at `http://localhost:8000/islandora/object/{mypid}/datastream/TN/download`. The results will be in `/tmp/drupal_debug.txt`.

# Additional Notes:

* Does this change require documentation to be updated? No, it includes an .api.php file for developers.
* Does this change add any new dependencies? No.
* Does this change require any other modifications to be made to the repository (ie. Regeneration activity, etc.)? No.
* Could this change impact execution of existing code? No.

# Interested parties
@bryjbrown, @Islandora/7-x-1-x-committers

### CODE REVIEW

**Comment by bryjbrown** [comment]
:+1: Just loaded it up and tested it, and it works as advertised. I left a few comments in the code review regarding wording, but other than that it looks good to go!

**Comment by mjordan** [comment]
@bryjbrown super, thanks!

**Comment by mjordan** [comment]
@bryjbrown sorry I can see your in-code comments - you said you left some but I want to make sure you didn't mean you were going to leaving some.

File: includes/utilities.inc [hunk 1]
```diff
@@ -44,6 +47,7 @@ function islandora_usage_stats_log_object_view(AbstractObject $object) {
         try {
           $collection_id = islandora_usage_stats_get_pid_id($collection);
           islandora_usage_stats_collection_view_to_db($object_access_id, $collection_id);
+          module_invoke_all('islandora_usage_stats_access_recorded', 'collection', $collection_id);
```
Comment (bryjbrown) [inline includes/utilities.inc]: I think this part can be a bit confusing, since it sounds like you are recording when collections are accessed, but what you are truly recording is the parent of the object being accessed. For this reason, I'd change the "$type" param from "collection" to "parent" to make it more explicit what is being recorded here.

File: islandora_usage_stats.api.php [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+ *
+ * @param string $type
+ *   One of 'object', 'collection', 'cmodel', 'ds'.
+ * @param int $access_id
+ *   The database ID resulting from persisting the entry to the database.
+ *   If the attempt to persist the record failed, this will be NULL.
+ * @param int $pid_id
+ *   For object access events, the ID of the row recording the event.
+ * @param string $pid
+ *   For object access events, the objec't PID.
```
Comment (bryjbrown) [inline islandora_usage_stats.api.php]: s/objec't/object's/

**Review by bryjbrown — CHANGES_REQUESTED** [review]
Looks good, just needs a few minor wording changes IMHO. The code works though.

**Comment by bryjbrown** [comment]
@mjordan Sorry about that, I made some comments on the code, but I don't think I formally saved it as a code review so you probably couldn't see it. I re-did it, hopefully you can see it this time.

File: islandora_usage_stats.api.php (hunk 2 above)
Comment (mjordan) [reply islandora_usage_stats.api.php]: Duh, will fix.

File: includes/utilities.inc (hunk 1 above)
Comment (mjordan) [reply includes/utilities.inc]: I see your point, but my intent was to make the values of `$type` consistent with the rest of the code, which uses 'collection'. I'll double check, but my understanding is that when an object is accessed, the hit is counted against the collection only, and not against anything else that happens to be the object's parent structurally; in other words, a hit is not recorded against a compound object when one if its child objects is accessed. Makes no difference to me which value we use though, as long as it's clear.

Maybe "parent_collection" is the best of both?

Comment (bryjbrown) [reply includes/utilities.inc]: @mjordan That sounds like a great idea to me, clearer names are always better.

Comment (mjordan) [reply includes/utilities.inc]: Great, I'll use that.

**Comment by mjordan** [comment]
Ah crap - I pushed a commit to origin (this repo) instead of mjordan. What should I do? Can we just delete the 7.x-ISLANDORA-2170 branch from https://github.com/Islandora/islandora_usage_stats?

**Comment by rosiel** [comment]
O no Mark what have you done??! Oh, it was just a separate branch? Phew. :) Yep - deleted.

**Comment by rosiel** [comment]
We can ignore "continuous-integration/travis-ci/push — The Travis CI build failed" - that was an automatically generated test on the accidental branch, islandora/7.x-ISLANDORA-2170. It succeeded on all boxes except one, as often happens. Now the branch doesn't exist any more, that test can't run. So long as "continuous-integration/travis-ci/pr" (ending in pr)  is happy, this PR is ok.

**Comment by mjordan** [comment]
@rosiel thanks, sorry about that.

File: includes/utilities.inc [hunk 3]
```diff
@@ -20,12 +20,15 @@ function islandora_usage_stats_log_object_view(AbstractObject $object) {
       $pid_id = islandora_usage_stats_get_pid_id($object->id);
       $object_access_id = islandora_usage_stats_object_view_to_db($pid_id);
 
+      module_invoke_all('islandora_usage_stats_access_recorded', 'object', $object_access_id, $pid_id, $object->id);
```
Comment (DiegoPino) [inline includes/utilities.inc]: Folks: @mjordan @bryjbrown wonder if making this new hook less generic could be something you would agree on? ```islandora_usage_stats_access_recorded``` right now  requires some conditionals (in their implementation) based on $type. That actually requires some extra care for implementers (i'm not one but I would assume its easy to avoid checking for all cases) like only taking care if $pid if passed.. and of type object, or collection, etc.. So what about A) splitting "object based" and "DS based" into 2 different ones, and passing $object instead of the pid for the first one. that means, if someone wants to record the models, well, it can be extracted from there but no need to invoke twice. Also, right now types are restricting something that actually does not need to be restricted at that level. What if someone wants to record a book access if a page was being seen? To avoid taking all "types" i account, maybe that could be left to the implementer by just passing the $object itself.

File: includes/utilities.inc [hunk 4]
```diff
@@ -20,12 +20,15 @@ function islandora_usage_stats_log_object_view(AbstractObject $object) {
...
+
       // Log content models.
       foreach ($object->models as $model) {
         if ($model != 'fedora-system:FedoraObject-3.0') {
           try {
             $model_id = islandora_usage_stats_get_pid_id($model);
             islandora_usage_stats_content_model_view_to_db($object_access_id, $model_id);
+            module_invoke_all('islandora_usage_stats_access_recorded', 'cmodel', $model_id);
```
Comment (DiegoPino) [inline includes/utilities.inc]: This here is confusing -could lead to issues-. The new hook is no CMODEL specific hook, so actually the same hooks will be invoked everytime for every CMODEL. And all will depend on how smart the logic inside those is to avoid making mistakes/recording too many times, or losing computing cycles (feel old when i say that.. i know..). See my previous comment.

**Review by DiegoPino — COMMENTED** [review]
Some comments on the new API hook here i would like you to consider. Let me know what you think. Won´t request changes formally because I'm not clear about the use cases which also leads to, for sake of future documentation, could you add the use cases to the pull request message 😄 . Thanks!

File: includes/utilities.inc (hunk 3 above)
Comment (mjordan) [reply includes/utilities.inc]: @DiegoPino thanks for comments.

> for sake of future documentation, could you add the use cases to the pull request message

Use cases are mentioned in the JIRA ticket, e.g. "By implementing this hook, modules could persist additional data such as referrer, or geolocation data for the event." I like the idea of supplying use case, but I wonder if the JIRA ticket is good enough for that.

> islandora_usage_stats_access_recorded right now requires some conditionals (in their implementation) based on $type.

hook_islandora_object_alter, which uses a $context that has numerous options within each sub-array. How is that any different than the hook definition in this PR? 

> Also, right now types are restricting something that actually does not need to be restricted at that level.

Those four types are the only ones that he Stats base module supports. I probably missing something, but how would not following that pattern in this hook definition get around that? I agree that passing the `$object` is always useful though, simply because you can do pretty much anything in the implementation.

File: includes/utilities.inc (hunk 4 above)
Comment (mjordan) [reply includes/utilities.inc]: Efficiency is important in this hook, so if providing a way to implement cmodel-specific implementations is more efficient, I can't argue with that. So you're suggesting something like defining `hook_cmodel_pid_islandora_usage_stats_access_recorded()`?

File: includes/utilities.inc (hunk 3 above)
Comment (DiegoPino) [reply includes/utilities.inc]: @mjordan thanks. So Sorry i missed the use case from the email (damn phones and their tiny screens)!
i get your point about hook_islandora_object_alter and *in no way i'm blocking this*,

I just feel there are too many invocations to the same set of hooks and for that, I felt 

A) having different hooks would make more sense, or, 

B) invoke them only once and actually pass a context. 

Because that hook(islandora_object_alter) is invoked a single time for a single PHP object type,
 In this case, the invocation is multiple in a single run and operating over two different (possible?) types of PHP Objects, `datastreams` and `Objects` letting (forcing) the implementer to discern when to actually do heavy work.

Sometimes it can be good to allow a hook  to be a little less smart because it will be only invoked under certain circumstances, imagine like prefiltered reality. I only say this because, yes, we have pretty smart hooks that deal with the full rainbow of scenarios in islandora too. So not a  must.

What i'm trying to say here is that your hook implementation will have to, at least, have something like this to work properly
```PHP
if (isset($pid) && ($type=='object)) { 
// deal with objects, but not cmodels which are objects, and  also not collections which are also objects) 
}
else { 
  if ($type!= `ds`) {
   //do datastream stuff 
  } 
  else {
    // do collection, cmodel stuff
   }
}
```

So basically that becomes a template. If that is the case and you all are OK with that, maybe adding that to the API hook body (the example) to let people understand what happens and what is needed to have a sane experience writing this hook could be a plus. Thanks!

File: includes/utilities.inc (hunk 4 above)
Comment (mjordan) [reply includes/utilities.inc]: @DiegoPino thinking about this a bit more, I want to clarify that this hook doesn't inspect the cmodel of the object (which is what other `hook_cmodel_pid_` hooks do), it fires when the base module records an access on the cmodel. Don't know if that's relevant to this discussion, just thought I'd mention it.

File: includes/utilities.inc (hunk 3 above)
Comment (mjordan) [reply includes/utilities.inc]: I'm OK with it, but also interested to hear what others say. I agree that the example code you provide here would be a useful addition to the `.api.php` file and am happy to add it.

File: includes/utilities.inc (hunk 4 above)
Comment (bryjbrown) [reply includes/utilities.inc]: I would agree with @DiegoPino that if there IS a way to broadcast a generic hook (for all cmodels) AND a hook that is specific for cmodels, that would be preferred in case someone wanted to implement a hook, but only when citations are viewed for example.

File: includes/utilities.inc (hunk 3 above)
Comment (bryjbrown) [reply includes/utilities.inc]: I think @DiegoPino has convinced me here, too. Having separate hooks fire for when objects are recorded vs datastreams (and perhaps even breaking it out further so that you can implement a hook that only fires when the MODS is viewed for example) would give implementers more options.

File: includes/utilities.inc (hunk 4 above)
Comment (mjordan) [reply includes/utilities.inc]: OK, but if the cmodel is just a string value that indicates what is being counted (that is all it is), does supplying cmdodel-specific hook definitions really matter? I just want to make sure that I'm not misunderstanding something here.

File: includes/utilities.inc (hunk 3 above)
Comment (mjordan) [reply includes/utilities.inc]: OK, let me propose something, in a little while.

File: includes/utilities.inc (hunk 4 above)
Comment (bryjbrown) [reply includes/utilities.inc]: Lets say that we want to record the browser used for everyone viewing books. In the current situation, we would need to implement a hook and check to see what cmodel it is, and if its a book THEN do some extra behavior to log the user's browser somewhere. This wouldn't take a lot of time, but it would fire for EVERY OBJECT. 

Alternatively, if there was an `access_recorded` hook that just fired when books were being recorded, you would shave that time off of having the hook fire for non-book objects.

Now, as I write this, I'm also wondering if this use case couldn't also be supported by hooking into `islandora_object_view` or something similiar, which I believe usage stats is hooking into. Maybe this IS overkill. Shrug.

Comment (DonRichards) [reply includes/utilities.inc]: I don't think that would be overkill.

**Comment by mjordan** [comment]
I'd like to propose the following, taking into account the comments so far:

1. we change the `module_invoke_all()` at https://github.com/Islandora/islandora_usage_stats/blob/7.x/includes/utilities.inc#L28 to accommodate specific content models, using the standard Islandora `cmodel_pid` pattern for hook names.
1. as parameters to the hook, we pass in a `$type` value (as now), the relevant database ID (as now) but add a `$context` parameter that can contain the object PID, the datastream DSID, etc. depending on what the value of `$type` is.

Am I missing anything?

I'll be in meetings for a while but want to see if this approach addresess everyone's comments.

**Comment by mjordan** [comment]
I've redone this PR to break out a separate hook for each of object access, collection access, content model access, and datastream downloads. Additionally, for content model access, there is both a generic hook and a cmodel-specific version. See the .api.php file for details.

Also, `islandora_usage_stats_object_view_to_db()` returns the id of the row added to record the object access, but the corresponding functions for content model, collection, and datastream downloads don't:

* islandora_usage_stats_content_model_view_to_db
* islandora_usage_stats_collection_view_to_db
* islandora_usage_stats_datastream_download_to_db

I have changed these functions so they now do return their respective row IDs, which lets us include the IDs in hook implemementations. I will update the PR template above to reflect this new change.

**Comment by mjordan** [comment]
Oh yeah, if you're using the test module at https://github.com/mjordan/islandora_2170_test, do a `git pull` on it first.

**Review by bryjbrown — APPROVED** [review]

**Comment by bryjbrown** [comment]
@mjordan Sorry for the delay, this fell off my radar. I've tested it and can verify that it works as advertised. All code review issues have been cleared. Am I good to merge? @Islandora/7-x-1-x-committers

**Comment by mjordan** [comment]
@bryjbrown cool, thanks, I'll close the JIRA issue.
